Sources/PackageCollections/PackageIndex.swift

```diff
@@ -0,0 +1,186 @@
+        }
+        
+        self.configurationStorage.update(index: .init(url: nil)) { result in
+            callback(result.tryMap { () })
+        }
+    }
+
+    func get(callback: @escaping (Result<Foundation.URL?, Error>) -> Void) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.configurationStorage.get { result in
+            callback(result.tryMap { $0.url })
+        }
+    }
+
+    func getPackageMetadata(
+        identity: PackageIdentity,
+        location: String?,
+        callback: @escaping (Result<PackageCollectionsModel.PackageMetadata, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { _, _ in
+            // TODO: call package index's get metadata API
+            fatalError("Not implemented: \(#function)")
+        }
+    }
+    
+    func findPackages(
+        _ query: String,
+        callback: @escaping (Result<PackageCollectionsModel.PackageSearchResult, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { _, _ in
+            // TODO: call package index's search API
+            fatalError("Not implemented: \(#function)")
+        }
+    }
+
+    func listPackages(
+        offset: Int,
+        limit: Int,
+        callback: @escaping (Result<PackageCollectionsModel.PaginatedPackageList, Error>) -> Void
+    ) {
+        self.runIfConfigured(callback: callback) { _, _ in
+            // TODO: cap `limit`
+            // TODO: call package index's list API
+            fatalError("Not implemented: \(#function)")
+        }
+    }
+
+    private func runIfConfigured<T>(
+        callback: @escaping (Result<T, Error>) -> Void,
+        closure: @escaping (Foundation.URL, (Result<T, Error>) -> Void) -> Void
+    ) {
+        let callback = self.makeAsync(callback)
+        
+        guard self.enabled else {
+            return callback(.failure(PackageIndexError.featureDisabled))
+        }
+        
+        self.get { result in
+            switch result {
+            case .failure(let error):
+                callback(.failure(error))
+            case .success(.none):
+                callback(.failure(PackageIndexError.notConfigured))
+            case .success(.some(let url)):
+                closure(url, callback)
+            }
+        }
+    }
+
+    private func makeAsync<T>(_ closure: @escaping (Result<T, Error>) -> Void) -> (Result<T, Error>) -> Void {
+        { result in self.callbackQueue.async { closure(result) } }
+    }
+}
+
+// MARK: - PackageMetadataProvider conformance
+
+extension PackageIndex: PackageMetadataProvider {
+    func get(
+        identity: PackageIdentity,
+        location: String,
+        callback: @escaping (Result<PackageCollectionsModel.PackageBasicMetadata, Error>, PackageMetadataProviderContext?) -> Void
```

Comment:
is PackageMetadataProviderContext used outside the successful case?

Reply:
Yes, see https://github.com/apple/swift-package-manager/pull/4010#discussion_r783556211